### PR TITLE
Fixes & code improvements for Mesh Nodes

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Made fixes & added code improvements in the new code handling AiMesh Nodes.

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

Here are a couple of screenshots showing some of the bugs fixed by this PR (I have no Mesh nodes at all).
The other bugs were around the "Email Notifications" menu options.

MeshNodes_MenuBug1

MeshNodes_MenuBug2

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,
Please run your own testing & verification to double-check this PR and make sure your mesh nodes are working as intended.
I did as much testing as I could but was limited because of the lack of actual mesh nodes.

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> Here are a couple of screenshots showing some of the bugs fixed by this PR (I have no Mesh nodes at all). The other bugs were around the "Email Notifications" menu options.
> 
> MeshNodes_MenuBug1
> 
> MeshNodes_MenuBug2

I haven't had a chance to sit down and review this yet, watching a movie, but why was it showing the menu for you?

**Comment by ExtremeFiretop** [comment]
Oh. LOL the node_list was being populated by the error message.
We can just removed it all together right? I don't see a reason to keep the error anymore that was when I was debugging the function.

I see you just made it clear the contents of node_list if it wasn't successful, I'm thinking trim the fat were possible.
I'll review more when I'm done the movie

**Comment by Martinski4GitHub** [comment]
> Oh. LOL the node_list was being populated by the error message. We can just removed it all together right? I don't see a reason to keep the error anymore that was when I was debugging the function.
> 
> I see you just made it clear the contents of node_list if it wasn't successful, I'm thinking trim the fat were possible. I'll review more when I'm done the movie
>
Yes, that's correct. The error message was taken as the "output" of the function. The proper handling in this case is to always check for the return code & adjust accordingly. This way it works no matter what, as long as the return code is correct & reflects the valid or invalid output. This is a very simple but effective technique, part of a larger "defensive programming" framework used in most languages, especially compiled executables.

File: MerlinAU.sh [hunk 1]
```diff
@@ -748,7 +748,7 @@ Get_Custom_Setting()
     if [ $# -eq 0 ] || [ -z "$1" ]; then echo "**ERROR**"; return 1; fi
     [ ! -d "$SETTINGS_DIR" ] && mkdir -m 755 -p "$SETTINGS_DIR"
 
-    local setting_value="" setting_type="$1" default_value="TBD"
+    local setting_value=""  setting_type="$1"  default_value="TBD"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:751]: Approved.

File: MerlinAU.sh [hunk 2]
```diff
@@ -782,16 +782,15 @@ Get_Custom_Setting()
...
-## Added by ExtremeFiretop [2024-Apr-02] ##
-##---------------------------------------##
-_GetAllNodeSettings_() {
-    if [ $# -lt 2 ]; then
-        return 1
-    fi
+##----------------------------------------##
+## Modified by Martinski W. [2024-Apr-06] ##
+##----------------------------------------##
+_GetAllNodeSettings_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:788]: Approved.

File: MerlinAU.sh [hunk 3]
```diff
@@ -836,8 +835,10 @@ Update_Custom_Settings()
         "FW_Auto_Backupmon" | \
         "FW_New_Update_Notification_Date" | \
         "FW_New_Update_Notification_Vers")
-            if [ -f "$SETTINGSFILE" ]; then
-                if [ "$(grep -c "$setting_type" "$SETTINGSFILE")" -gt 0 ]; then
+            if [ -f "$SETTINGSFILE" ]
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:838]: Approved.

File: MerlinAU.sh [hunk 4]
```diff
@@ -2204,56 +2204,66 @@ _GetLoginCredentials_()
...
 
-##---------------------------------------##
-## Added by ExtremeFiretop [2024-Mar-26] ##
-##---------------------------------------##
-_PermNodeList_() {
-    # Get the value of asus_device_list
+##----------------------------------------##
+## Modified by Martinski W. [2024-Apr-06] ##
+##----------------------------------------##
+_GetNodeIPv4List_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2210]: Approved.

Comment (ExtremeFiretop) [reply MerlinAU.sh:2210]: May remove some of the echo error messages from the functions. Will see...

File: MerlinAU.sh [hunk 5]
```diff
@@ -2204,56 +2204,66 @@ _GetLoginCredentials_()
...
 
-##---------------------------------------##
-## Added by ExtremeFiretop [2024-Mar-27] ##
-##---------------------------------------##
-_NodeActiveStatus_() {
-    # Get the value of cfg_device_list
+##----------------------------------------##
+## Modified by Martinski W. [2024-Apr-06] ##
+##----------------------------------------##
+_NodeActiveStatus_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2241]: Same with this, approved formatting improvements, may still go back to remove the echo error messages.

File: MerlinAU.sh [hunk 6]
```diff
@@ -2348,7 +2358,7 @@ _GetNodeInfo_()
     fi
 
     # Run the curl command to retrieve the HTML content
-    htmlContent=$(curl -s -k "${NodeURLstr}/appGet.cgi?hook=nvram_get(productid)%3bnvram_get(asus_device_list)%3bnvram_get(cfg_device_list)%3bnvram_get(firmver)%3bnvram_get(buildno)%3bnvram_get(extendno)%3bnvram_get(webs_state_flag)%3bnvram_get(odmpid)%3bnvram_get(wps_modelnum)%3bnvram_get(model)%3bnvram_get(build_name)%3bnvram_get(lan_hostname)%3bnvram_get(webs_state_info)%3bnvram_get(label_mac)" \
+    htmlContent="$(curl -s -k "${NodeURLstr}/appGet.cgi?hook=nvram_get(productid)%3bnvram_get(asus_device_list)%3bnvram_get(cfg_device_list)%3bnvram_get(firmver)%3bnvram_get(buildno)%3bnvram_get(extendno)%3bnvram_get(webs_state_flag)%3bnvram_get(odmpid)%3bnvram_get(wps_modelnum)%3bnvram_get(model)%3bnvram_get(build_name)%3bnvram_get(lan_hostname)%3bnvram_get(webs_state_info)%3bnvram_get(label_mac)" \
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2361]: Approved.

File: MerlinAU.sh [hunk 7]
```diff
@@ -2357,27 +2367,27 @@ _GetNodeInfo_()
...
-    node_extendno=$(echo "$htmlContent" | grep -o '"extendno":"[^"]*' | sed 's/"extendno":"//')
-    node_webs_state_flag=$(echo "$htmlContent" | grep -o '"webs_state_flag":"[^"]*' | sed 's/"webs_state_flag":"//')
-    node_webs_state_info=$(echo "$htmlContent" | grep -o '"webs_state_info":"[^"]*' | sed 's/"webs_state_info":"//')
-    node_odmpid=$(echo "$htmlContent" | grep -o '"odmpid":"[^"]*' | sed 's/"odmpid":"//')
-    node_wps_modelnum=$(echo "$htmlContent" | grep -o '"wps_modelnum":"[^"]*' | sed 's/"wps_modelnum":"//')
-    node_model=$(echo "$htmlContent" | grep -o '"model":"[^"]*' | sed 's/"model":"//')
-    node_build_name=$(echo "$htmlContent" | grep -o '"build_name":"[^"]*' | sed 's/"build_name":"//')
-    node_lan_hostname=$(echo "$htmlContent" | grep -o '"lan_hostname":"[^"]*' | sed 's/"lan_hostname":"//')
-    node_label_mac=$(echo "$htmlContent" | grep -o '"label_mac":"[^"]*' | sed 's/"label_mac":"//')
+    # Extract values using regular expressions #
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2376]: Approved.

File: MerlinAU.sh [hunk 8]
```diff
@@ -2399,21 +2409,29 @@ _GetNodeInfo_()
     fi
 }
 
-##------------------------------------------##
-## Modified by ExtremeFiretop [2024-Mar-27] ##
-##--------------------------------------- --##
+##----------------------------------------##
+## Modified by Martinski W. [2024-Apr-06] ##
+##----------------------------------------##
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2414]: Approved extra defensive checks

File: MerlinAU.sh [hunk 9]
```diff
@@ -4366,7 +4384,8 @@ _SimpleNotificationDate_()
 ## Added by ExtremeFiretop [2024-Mar-27] ##
 ##---------------------------------------##
 # Define a function to print information about each AiMesh node
-_PrintNodeInfo() {
+_PrintNodeInfo()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:4387]: Approved formatting improvements.

File: MerlinAU.sh [hunk 10]
```diff
@@ -4592,28 +4612,33 @@ _ShowAdvancedOptionsMenu_()
    fi
 
    # Additional Email Notification Options #
-   if _CheckEMailConfigFileFromAMTM_ 0 && "$sendEMailNotificationsFlag"
+   if _CheckEMailConfigFileFromAMTM_ 0
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:4615]: Approved bug fix!

File: MerlinAU.sh [hunk 11]
```diff
@@ -4641,32 +4666,57 @@ _ShowAdvancedOptionsMenu_()
...
-   then echo "**ERROR** **NO_PARAMS**" ; return 1 ; fi
+# RegExp for IPv4 address #
+readonly IPv4octet_RegEx="([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-5][0-5])"
+readonly IPv4addrs_RegEx="((${IPv4octet_RegEx}\.){3}${IPv4octet_RegEx})"
+readonly IPv4privt_RegEx="((^10\.)|(^172\.1[6-9]\.)|(^172\.2[0-9]\.)|(^172\.3[0-1]\.)|(^192\.168\.))"
+
+##----------------------------------------##
+## Modified by Martinski W. [2024-Apr-06] ##
+##----------------------------------------##
+_ValidatePrivateIPv4Address_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:4677]: Approved new function to validate IPs

**Review by ExtremeFiretop — APPROVED** [review]

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

So far, everything looks good! Will test a bit more and advise!
Will probably trim some fat I don't think we need anymore, also will sync up with the Gnuton branch so it has these changes :)

**Comment by Martinski4GitHub** [comment]
> @Martinski4GitHub
> 
> So far, everything looks good! Will test a bit more and advise! Will probably trim some fat I don't think we need anymore, also will sync up with the Gnuton branch so it has these changes :)
>
OK, sounds good. I'll be up a couple of more hours before going to bed so for any questions just let me know.

**Pull request merged**